fixed a lot of hacky stuff

alexanderankin opened the pull request:
going to continue this progress through the summer most likely, mostly a pull request to draw attention, because when i started working with/on this library it looked like it was out of maintenance.

checkout the code coverage badge.

RomanKharin commented:
Wow, that's huge.
More than can be reviewed by one man.
It should be splitted to several pull requests.
Some thoughts:
1. change formatting - must have
2. tox - too
3. new tests: no binary pdf in repo (only hash), no external http depencies without hash
4. attic is useful, there are few ideas inside
5. why you delete all out tests and docs?

alexanderankin commented:
i was thinking about doing pdfs as hashes but decided against it, because really the library produces different pdf's, because it bakes `/CreationDate` pdf flag into the document, i wrote a utility function that compares binary files ignoring this detail. So this is done for precision.

I deleted everything that isnt formatted or organized and am slowly organizing. I deleted all test files, am rewriting the tests as test cases some are already done. (code coverage is 50%, i wrote 0 of those tests).

I hope to include sphinx docs soon, because there are some irregularities (inevitable ones) when documentation is separate md files, and some of the methods return some weird stuff.

alexanderankin commented:
though i would ideally be pulling code back in from reinghart and making merging easier in the future, my immediate top priorities are
* porting all of the tests into actual tests and not like example code,
  * making test suite rigorous enough to comfortably play around with core class (what a stinking pile, half the writer methods can be totally abstracted into another class, maybe inheritance maybe containerization)
* pulling down the features that you guys have with image masking and stuff, and 
* making table functionality core (this is my actual application)

i welcome any additional input or specific requests or concerns you guys may have for me! the more concrete the orders are the better and more likely i can cater to them.

RomanKharin commented:
> because really the library produces different pdf's, because it bakes /CreationDate pdf flag into the document
It can be avoided with this tip
```pdf._putinfo = lambda: common.test_putinfo(pdf)```
Anyway currently we support custom test suite and unittest (sort of) with single codebase.

Let's change things slightly.  Main plan there https://github.com/reingart/pyfpdf/wiki/Roadmap

alexanderankin commented:
hey so according to some of the text in the pdf spec, pdf/a just has something to do with putting all the important stuff in the file in the beginning of the file to optimize random access on disk or off of a wire. so i think that can totally be simply lifted off a different library (maybe not quite as extreme as calling a shell command or something, but you get the point, I dont see why this needs to be closely integrated, although I do admit this might be in ignorance of something).

I havent copied any tests regarding fonts yet, so maybe you can shed some light on what `Support for 64k+ font ranges` means and i can try and work on that. As far as all of the stuff under 2.0, hopefully my organizational efforts can enable more efficient and developer friendly yada yada to help someone who knows about pdf get that done.

pep-8 is proving to be a real doozy because although fpdf.py module itself looks better now, I have eight or so rule exceptions, and flake8 is really uncompromising where the pep8 standard does not insist on certain things, like hanging indents, et cetera. For this reason I have not invested the extra time into cleaning up the other modules, but you can see the code in tox.ini and the other tests in tox.ini pass (feel free to clone my master branch locally and look around.

I have also yet to copy over the readthedocs files because I dont know how those files get uploaded but I'd insist on having the script that processes those exist alongside, and having not found one, i sort of cleaned them up as orphans in the repository.

that aside, my next steps are probably going to be breaking up the fpdf class into FPDF and ShapeWriter, DocumentWriter, HTMLWriter, ImageWriter, FontWriter, because the tests that exist are not organized and this would be a great way to organize the features, the tests for them, and the documentation? Not sure if this is a good idea, or which would come first, the tests or the new Classes, would love to hear your input on this. Also, in a major version change, this could come with a usage like so: the user would import whatever is needed, and create a class that inherits from all of these classes:

```py
class Document(fpdf.FPDF, fpdf.FontWriter, fpdf.DocumentWriter):
  pass

d = Document()
d.add_page()
d.set_font('Arial', size=12)
d.cell(w=0, txt="hello world")
d.output("hello_world.pdf")
```

also a feature I thought that would be really useful to me is having the height parameter in cell and multicell default to a 'line spacing' instance parameter which would be MSWord default: 1.15 * font size in points; other improvements in that same vein which crossed my mind are: removing all default parameters, adding an option to cell function to not write text past a designated width (currently width only affects border drawing).

More comments while i am writing all of this down, about the roadmap plans: generating pdf objects will need to have very careful tests written to ensure that the spirit of "calling functions in order" that this library has over other libraries be preserved.

look forward to your thoughts!

renovate[bot] committed:
Update github/codeql-action action to v3.30.5 (#1587)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
Store bidi reference files locally (#1584)

* store bidi reference files locally

* fix to save as binary

andersonhc committed:
Fix typo on Text.md (#1590)

andersonhc committed:
fix typos (#1591)

renovate[bot] committed:
Update crate-ci/typos action to v1.37.0 (#1589)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

tanmayc07 committed:
doc: add fpdf2 usage with bottle web framework (#1592)

allcontributors[bot] committed:
add tanmayc07 as a contributor for doc (#1593)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.37.1 (#1594)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v3.30.6 (#1596)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.37.2 (#1598)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add Malay-dev as a contributor for code (#1599)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

Malay-dev committed:
Feat: Add support for unicode_range in add_font()  (#1595)

* Feat: support unicode_range in add_font()

* Feat: Add test cases for unicode_range

* Feat: Add required test pdfs and font for unit testing

* Update docstring for unicode_range param

* Update CHANGELOG for unicode_range param support

* Update CHANGELOG for unicode_range param support

* Feat: refactor support for unicode_range in add_font()

* doc: add unicode_range param usage

* Fix: test for unicode_range specialized font for punctuation

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.38.0 (#1600)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v3.30.7 (#1604)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4 (#1605)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.38.1 (#1606)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.30.8 (#1608)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add Rapteon as a contributor for code (#1609)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

Lucas-C committed:
Docs: minor WSGI section reorg (#1612)

tanmayc07 committed:
doc: add fpdf2 usage with cherrypy web framework (#1611)

* doc: add fpdf2 usage with cherrypy web framework

* docs: update cherrypy description

Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

* docs: apply suggestion from @Lucas-C

Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

---------

Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

Vidushi2709 committed:
docs: Add Plone usage documentation to UsageInWebAPI.md (#1610)

* docs: Add Plone usage documentation to UsageInWebAPI.md

* docs

allcontributors[bot] committed:
add Vidushi2709 as a contributor for doc (#1614)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

Rapteon committed:
(feat): Added optional parameter `variations` in add_font (#1607)

* (feat): Added optional parameter `variations` in add_font
See #1544 for details.

* (feat): Added optional parameter `variations` for add_font
Made changes suggested in the review. See #1544 for more details and PR #1607 for review.

* (feat): Added optional parameter `variations` in add_font
Generated pdf to fix failing test.

* (feat): Added optional parameter `variations` in add_font
Added section for variable fonts in documentation.

andersonhc committed:
Simplified SVG <text> tag rendering (#1601)

* drawing/svg text implementation

* add TextRun

* add new tests and changelog entry

allcontributors[bot] committed:
add Rapteon as a contributor for ideas (#1617)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

Lucas-C committed:
Minor documentation fixups (#1618)

Lucas-C committed:
Docs: minor improvements (#1619)

Lucas-C committed:
Dropping support for Python 3.9 (#1620)

andersonhc committed:
Add Python 3.14 support (#1622)

* add python 3.14 support

* add changelog message

* Update test/image/test_image_info.py

Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

* Update .github/workflows/continuous-integration-workflow.yml

Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

* temporarily restrict Pillow version

* add comments

---------

Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update anchore/scan-action action to v7.0.1 (#1623)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
remove dependency on requirements.txt (#1625)

renovate[bot] committed:
chore(deps): update anchore/scan-action action to v7.0.2 (#1626)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Vedant817 committed:
add palette parameter to add_font() for color font palette selection (#1621)

* feat: add palette parameter to add_font() method for color font palette selection

* feat: add palette_index support to TTFFont class for color font palettes

* feat: implement configurable palette selection in COLRFont class

* test: add comprehensive test script for palette selection feature

* test: add unit tests for color font palette selection feature

* docs: add usage example for color font palette selection

* Add documentation for color font palette selection feature

- Include usage example and explanation in docs/Unicode.md for the palette parameter in add_font().

* Improve palette selection tests with proper assertions and golden PDFs

- Update test_palette_selection.py: Replace generic assertions with assert_pdf_equal, add tmp_path/caplog fixtures, and verify warning logs for out-of-range palettes.
- Add golden PDF files for reliable test output comparisons and validation.

* Fix lint issues and formatting for palette selection PR

- Auto-format code with Black across modified files.
- Make scripts/create-endesive-certificates.sh executable.
- Remove commented print statements from fpdf/drawing.py.
- Install pylint and ensure pre-commit hooks pass.

* Fixed the linting error for the font_type_3.py

* fix: resolve pre-commit failures and code quality issues

- Fix pylint warnings by removing useless suppressions and adding proper ones
- Update CHANGELOG.md with resolved issues
- Ensure all source files achieve 10.00/10 pylint score

* update CHANGELOG.md: document palette parameter support for color font selection

allcontributors[bot] committed:
add Vedant817 as a contributor for code (#1627)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

andersonhc committed:
Update version number for color font palette feature (#1628)

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.30.9 (#1629)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
Pillow v12 support (#1631)

allcontributors[bot] committed:
add alhman as a contributor for bug (#1633)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add Ai-Chetan as a contributor for code (#1635)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

andersonhc committed:
COLRv1 support enhancements (#1636)

* gradient spread methods and sweep gradients

add test for sweep gradients and gradient spread

add blending groups

* implement colrv1 clipping boxes

* fix color font glyph object compress

* add changelog

* black 25.9.0 changes

renovate[bot] committed:
chore(deps): update dependency python to 3.14 (#1602)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
fix comments now that lint job is running on python 3.14 (#1638)

allcontributors[bot] committed:
add ystepanoff as a contributor for translation (#1639)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

andersonhc committed:
revert lint job back to python 3.13 due to guarddog/pygit2 issues (#1640)

ystepanoff committed:
Add tutorial page in Ukrainian (#1637)

* Add tutorial page in Ukrainian

* Add changelog note

* Fix the changelog entry

* Add a link to the tutorial in Ukrainian to index.md

* Apply suggestion from @andersonhc

Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>

* Update CHANGELOG.md

Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

---------

Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>
Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update anchore/scan-action action to v7.1.0 (#1643)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Ai-Chetan committed:
Fix rounded rectangle rendering with thick borders and fill artifacts (#1634)

* Fix rounded rectangle rendering with thick borders and fill artifacts

* Address review feedback: remove validation script, rename test file, update docstrings, run black and pylint

* Add tests for round corners with style='D' (draw only) and style='F' (fill only)

* docs: add changelog entry for rounded-rect fix (#1634)

andersonhc committed:
review changelog (#1645)

andersonhc committed:
Create parameter deprecation annotation (#1642)

* change parameter deprecation

* remove assert on record length

* change test

* add changelog

* add deprecated since information

* remove invalid assert

* add deprecation policy on Development.md

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.31.0 (#1648)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

RaphaelFLT committed:
Migrate project metadata from setup.py to pyproject.toml (PEP 639 Compliance) (#1615)

* migration du projet pour l'utilisation de pyproject.toml

* Correction License to the right format

* Update pyproject.toml

Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

* fix: Resolve maintainer review comments regarding version variables.

* fix:small correction in fpdf.py requested during review, moved __version__ after FPDF_VERSION

* fix : addded fonttools

* style: Apply black formatting

* fix: Correct dynamic version configuration in pyproject.toml

---------

Co-authored-by: FRELAUT <rafflt@MacBook-Air-de-Raphael.local>
Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

allcontributors[bot] committed:
add RaphaelFLT as a contributor for code (#1650)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

Lucas-C committed:
fix CI pipeline & new release 2.8.5 (#1651)

* fix CI pipeline & new release 2.8.5

* fix min Pillow version

---------

Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>

Lucas-C committed:
Fix publishing of new release 2.8.5 (#1652)

andersonhc committed:
change build command (#1653)

andersonhc committed:
mkdocs step: change setup.py reference (#1654)

* mkdocs step: change setup.py reference

* install build for the publish step

Lucas-C committed:
Fix license metadata in pyproject.toml (#1655)

Lucas-C committed:
Fix packaging error due to use of pkg_resources (#1656)

Lucas-C committed:
Fix CI pipeline for Python 3.10 with old dependencies (#1657)

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.31.2 (#1658)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.39.0 (#1659)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.13.2 (#1661)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.39.1 (#1663)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add Ani07-05 as a contributor for code (#1664)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

Ani07-05 committed:
Feature/svg gradient support (#1647)

* Add SVG gradient support (linear and radial)

Fixes #1580

Features:
- Linear gradients with all SVG coordinate systems
- Radial gradients with focal point (fx, fy, fr) support
- Gradient stops with color and opacity
- Spread methods: pad, reflect, repeat (related to PR #1636)
- Gradient transforms (rotation, scaling, translation)
- Support for both objectBoundingBox and userSpaceOnUse units
- Gradients can be applied to both fill and stroke

Implementation:
- Added gradient parsing methods in SVGObject class
- Integrated with existing fpdf2 GradientPaint system
- Gradients stored in gradient_definitions dict by ID
- Applied via url(#gradientId) references in fill/stroke

Tests:
- 18 comprehensive test cases covering all gradient features
- All 287 SVG tests passing (18 new + 269 existing)
- Added 5 SVG test fixtures and 5 PDF reference files

* docs: Add SVG gradient documentation and CHANGELOG entry

- Added gradient support section to docs/SVG.md with examples
- Removed gradients from unsupported features list
- Added CHANGELOG entry for gradient support

* fix: Address code review feedback

- Fix logging f-string warnings - use % formatting instead
- Add spread_method parameter to GradientPaint constructors
- Refactor tests to plain functions, remove class wrapper
- Clean up test imports
- Add gradient SVGs to integration tests

* Fix pylint errors, update failing test, and address PR review feedback

- Convert _convert_gradient_coordinate, _parse_gradient_stops, and
  _extract_gradient_id to static methods (fixes R6301: no-self-use)
- Replace broad Exception catches with specific exception types
  (ValueError, AttributeError, TypeError) on lines 1024 and 1102
- Fix test_svg_image_with_unsupported_color to test HSL colors instead
  of gradients (which are now supported)
- Remove detailed Gradients subsection from SVG.md per reviewer request
- Create [2.8.6] section in CHANGELOG.md and move SVG gradient entry

andersonhc committed:
Change unsupported color test (#1665)

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.39.2 (#1666)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.31.3 (#1667)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Lucas-C committed:
Fix bug in tables borders rendering - close #1669 (#1670)

renovate[bot] committed:
chore(deps): update actions/checkout action to v5.0.1 (#1672)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update anchore/scan-action action to v7.2.0 (#1673)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.31.4 (#1674)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add philj56 as a contributor for code (#1676)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

philj56 committed:
Improve image performance (#1675)

* Don't unnecessarily calculate bounding boxes.

This greatly speeds up path drawing when gradients aren't used.

* Use slot names directly when merging styles.

This speeds up merging `GraphicsStyle`s by skipping various correctness
checks and extra getattr/setattr calls.

* Use deepcopy when merging styles.

This slightly speeds up merging.

* Prevent unnecessary path copies when drawing SVG.

* Skip setters in GraphicsStyle's __init__().

* Speed up GraphicStyle's __setattr__ slightly.

* Simplify and speed up image alpha testing.

* Optionally use numpy to speed up alpha testing.

This mainly reduces poor performance for opaque images with an alpha
channel.

* Remove underscore from public API parameter.

* Add performance improvement changes to CHANGELOG.

---------

Co-authored-by: Philip Jones <philip.jones@cellularhighways.com>
Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update actions/checkout action to v6 (#1677)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.31.5 (#1679)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update anchore/scan-action action to v7.2.1 (#1681)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update actions/setup-python action to v6.1.0 (#1682)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Lucas-C committed:
Moving from requirements.txt to pyproject.toml dependency-groups (#1671)

* Moving from requirements.txt to pyproject.toml dependency-groups - close #1613

* change to project.optional-dependencies

* fix lint job dependency and add missing dependency on doc group

* create script to build dependency list for guarddog

* fix guarddog setup

* Final fixups

---------

Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.40.0 (#1683)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.31.6 (#1686)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.13.3 (#1687)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update actions/checkout action to v6.0.1 (#1688)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.31.7 (#1690)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

ByZay committed:
Translate of Tutorial 7 in French Version (#1684)

Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

allcontributors[bot] committed:
add ByZay as a contributor for doc (#1692)

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

Lucas-C committed:
Fix documentation building with mkdocs (#1693)

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.14.0 (#1695)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update anchore/scan-action action to v7.2.2 (#1696)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.31.8 (#1697)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.31.9 (#1701)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add bruceoberg as a contributor for bug (#1705)

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

Lucas-C committed:
A5 page dimensions fix #1699 (#1703)

Lucas-C committed:
Minor documentation fixup - close #1685 (#1707)

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.40.1 (#1708)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.41.0 (#1709)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
fix borb logo file path (#1711)

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.42.0 (#1710)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
Implement strict type hinting (#1691)

* typing

* add pyright precommit and runner step

* fix pyright errors

* remove some casts

* has_alpha optimization

* remove cast/assert for performance

* documentation and changelog

andersonhc committed:
typing fix (#1712)

* typing fix

* please pyright

renovate[bot] committed:
chore(deps): update anchore/scan-action action to v7.2.3 (#1713)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

slkoo-cc committed:
Add position image code in docs (#1714)

Co-authored-by: Lucas Cimon <925560+Lucas-C@users.noreply.github.com>

allcontributors[bot] committed:
add slkoo-cc as a contributor for doc (#1715)

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

Lucas-C committed:
Fix typo in docs/Images.md

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.31.10 (#1716)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
Reformat code with Black v26 (#1719)

* add .venv to gitignore

* changes to comply with black 26

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.42.1 (#1718)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
refactor GraphicsStateMixin (#1720)

* refactor GraphicsStateMixin

renovate[bot] committed:
chore(deps): update anchore/scan-action action to v7.3.0 (#1722)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update actions/setup-python action to v6.2.0 (#1723)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update actions/checkout action to v6.0.2 (#1724)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.31.11 (#1725)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.14.1 (#1726)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.42.2 (#1727)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.32.0 (#1728)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

BharathPESU committed:
Support WOFF and WOFF2 fonts (#1660)

* issue fixed

* Fix: optional imports and generator contextmanager typing in fpdf.py

* Fix type checking and performance issues in fonts.py and fpdf.py

- fpdf/fonts.py:
  * Added explanatory comment for TYPE_CHECKING block clarifying it's for static type checkers only
  * Optimized hbfont property to avoid performance hit for non-WOFF fonts by checking file extension
  * For WOFF/WOFF2: uses byte buffer decompression (required for HarfBuzz)
  * For TTF/OTF: loads directly from file path (faster, no extra serialization)
  * Removed invalid fallback for WOFF/WOFF2 that would fail anyway
  * Added logging for failed temp file cleanup to track orphaned files

- fpdf/fpdf.py:
  * Removed unnecessary TYPE_CHECKING block for optional dependencies
  * Fixed table() method return type from Iterator[Table] to ContextManager[Table]

* support woff and woff2

* copy is_compressed attribute

* docs and changelog

* remove some comments

---------

Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>

renovate[bot] committed:
chore(deps): update anchore/scan-action action to v7.3.1 (#1730)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.42.3 (#1731)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
Increase font compatibility: Font collections, CDI-keyed fonts and Microsoft symbol fonts (#1732)

* font compatibility changes

andersonhc committed:
Standardize float comparisons with FloatTolerance (#1733)

* font compatibility changes

* implement new float tolerance class

* fix undesired font inclusion

* add changelog

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.32.1 (#1735)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
Ignore invalid typo error (#1737)

* update typos pre-commit version

* fix typos

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.43.0 (#1736)

* chore(deps): update crate-ci/typos action to v1.43.0

RaphaelFLT committed:
Add FPDF.transform context manager (#1717)

This new method allows applying affine transformations (translation, rotation, scaling) to a block of code using drawing_primitives.Transform. It automatically handles the conversion between FPDF user units and PDF device units (including Y-axis inversion).

Co-authored-by: FRELAUT <rafflt@MacBook-Air-de-Raphael.local>

allcontributors[bot] committed:
add deltragon as a contributor for bug, and code (#1740)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

deltragon committed:
escape unbalanced parentheses (#1734)

* add tests for files with parentheses

* syntax: escape unbalanced parentheses

* changelog

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.43.1 (#1739)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update anchore/scan-action action to v7.3.2 (#1742)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.32.2 (#1743)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.43.2 (#1744)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.43.3 (#1745)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.14.2 (#1746)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
Add support for bitmap fonts (#1747)

* implement bitmap font rendering

Lucas-C committed:
[CI] Fix borb SVG logo rendering in docs (#1749)

andersonhc committed:
Refactor rotation(), skew() and mirror() methods to use the new transform() context (#1750)

* switch methods to new transform context manager

* changelog

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.43.4 (#1751)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.32.3 (#1753)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add amidou-naba as a contributor for code (#1754)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

amidou-naba committed:
Fix markdown character escaping bug (issue #1236) (#1741)

* Bug #1236 has been fixed

* fix lint issues

* last fixes

* Bug #1236 has been fixed

fix lint issues

last fixes

* improve tests

---------

Co-authored-by: Amidou Nabassaga <anabassaga@emi.u-bordeaux.fr>
Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>

andersonhc committed:
Refactor image info (#1721)

* refactor image info as TypedDict

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.43.5 (#1755)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
release 2.8.6 (#1756)

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.32.4 (#1758)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

srittau committed:
Fix a few typing issues (#1757)

* Fix a few typing issues

allcontributors[bot] committed:
add srittau as a contributor for code (#1759)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.15.0 (#1762)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
fix ebdt font detection (#1763)

paulsuh committed:
Add missing files to docs nav tree (#1766)

* Added missing links to docs nav tree

Tutorial-ua.md, NamedDestinations.md, DatabaseStorage.md

* Update CHANGELOG.md

allcontributors[bot] committed:
add paulsuh as a contributor for doc (#1767)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.44.0 (#1768)

* chore(deps): update crate-ci/typos action to v1.44.0

* ignore Suh on md files

---------

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>

andersonhc committed:
Revert "Refactor image info (#1721)" (#1769)

This reverts commit ae342be683c1a816f947de671cc6fa47742934cf.

* chore(deps): update crate-ci/typos action to v1.44.0

* ignore Suh on md files

---------

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
release v2.8.7 (#1770)

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.32.5 (#1771)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.32.6 (#1773)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.15.1 (#1774)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add CoLa5 as a contributor for bug (#1776)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

andersonhc committed:
Fix fallback-first font promotion on new pages (#1775)

* Fix problem with fallback font being first on page

* add changelog entry

andersonhc committed:
Fix tests for all python implementation using zlib-ng (#1778)

* fix tests for zlib-ng

* add changelog

andersonhc committed:
preserve bn characters on bidi (#1780)

Lucas-C committed:
Minor type annotation fixup (#1781)

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.16.0 (#1782)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.33.0 (#1783)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
Fix transform application on user space gradients (#1787)

* fix transform on user space gradients

* add changelog

* regenerate test without zlib-ng

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.34.0 (#1789)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.34.1 (#1790)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update anchore/scan-action action to v7.4.0 (#1791)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.35.1 (#1796)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
fix use of image tempfile (#1798)

allcontributors[bot] committed:
add makapuf as a contributor for bug (#1799)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.16.1 (#1800)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
mypy 1.20 (#1806)

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.45.0 (#1801)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

BillionClaw committed:
fix: remove obsolete [base] and [full] extras from test dependencies (#1795)

* fix: remove obsolete [base] and [full] extras from test dependencies

* Update CHANGELOG

Added dependency extras for camelot-py and endesive.

* Update CHANGELOG with new dependency extras

Added dependency extras for camelot-py and endesive in pyproject.toml.

---------

Co-authored-by: BillionClaw <267901332+BillionClaw@users.noreply.github.com>
Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>

renovate[bot] committed:
chore(deps): update pypa/gh-action-pypi-publish action to v1.14.0 (#1809)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.17.0 (#1810)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Pawansingh3889 committed:
docs: add Punjabi (pa) tutorial translation (#1805)

* docs: add Punjabi (pa) tutorial translation

* docs: add Punjabi tutorial to index, mkdocs nav, and changelog

---------

Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>

allcontributors[bot] committed:
add Pawansingh3889 as a contributor for translation (#1811)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add CoLa5 as a contributor for bug, and code (#1812)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

CoLa5 committed:
Preserve link annotations during `FPDF.multi_cell(..., dry_run=True)` (#1808)

* add tests for dry run link annotations bug

* Fix dry run link annotations bug

* note fix in change log

* fix position of text on pdf page

andersonhc committed:
update tests (#1813)

CoLa5 committed:
Fixes "Consecutive links in `FPDF.multi_cell(...)`" (#1815)

* Fix issue of consecutive md links do not split into seperate fragments

* Fix links to own github page

* Add doc of fix to changelog

renovate[bot] committed:
chore(deps): update crate-ci/typos action to v1.45.1 (#1817)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.18.0 (#1818)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update github/codeql-action action to v4.35.2 (#1821)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update step-security/harden-runner action to v2.19.0 (#1825)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
Add configurable resource access policy for image loading (#1824)

* implement resource load policy

* docs

* complete implementation

* cleanup

andersonhc committed:
typing fixes for pyright 1.1.409 (#1828)

* typing fixes for pyright 1.1.409

* fix typing for python 3.11-

renovate[bot] committed:
Update crate-ci/typos action to v1.45.2 (#1830)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

CoLa5 committed:
Support of markdown styled links (#1827)

* add support of styled markdown links

* test: Fix link of test_multi_cell_markdown_styled_link to github

* chore: Add fixed bug to changelog

* docs: Add strikethrough to markdown style

* docs: Add styling of md links to docs

* test: Fix html link underline test file to correctly be blue underlined although followed by black text

renovate[bot] committed:
Update crate-ci/typos action to v1.46.0 (#1833)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action action to v4.35.3 (#1834)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update step-security/harden-runner action to v2.19.1 (#1835)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

CoLa5 committed:
Fix bug of `FPDF.multi_cell` dry run in toc (#1838)

* Fix bug of `multi_cell(..., dry run=True)` inside of toc provoking new page, but internal counter of inserted pages in toc is not reset (`FPDF._toc_inserted_pages`) after dry run

* test(outline): Add height calculation as motivation

* docs: Update changelog

allcontributors[bot] committed:
add HippolyteCediey as a contributor for bug (#1842)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

andersonhc committed:
fix svg arc calculation (#1841)

renovate[bot] committed:
Update github/codeql-action action to v4.35.4 (#1845)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update crate-ci/typos action to v1.46.1 (#1846)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

CoLa5 committed:
Fix `FPDF.multi_cell(..., dry_run=True, output="LINES") to return markdown text (#1844)

* feat(fpdf.FPDF): Dry-run of `multi_cell` with `markdown=True` returns another markdown string to keep styling consistent

* fix: Simplify `_join_text_lines`

* docs: Update changelog

* fix(text): Keep escaped markdown markers

* test(text): Test keeping escaped markdown markers in `output="LINES"`

* fix(text): Fix bug of underline marker for link

* test(text): Test all markdown markers are returned in case of `output="LINES"`

* fix(text): Not escape markdown markers inside of links

renovate[bot] committed:
Update peaceiris/actions-gh-pages action to v4.1.0 (#1848)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

CoLa5 committed:
Fix ToC-rendering does not handle gstate (#1839)

* fix(toc): Fix last gstate leaking into toc

* test(toc): Test fix of last gstate leaking into toc

* docs: Update changelog

* fix(html): Fix html toc

* fix(outline): Fix gstate manager around ToC

* fix(outline): Fix footer style setting own font family and size in test

---------

Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>

renovate[bot] committed:
Update step-security/harden-runner action to v2.19.2 (#1850)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update step-security/harden-runner action to v2.19.3 (#1851)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Theo1335 committed:
Add basic <symbol> support in SVG  (#1819)

* Automatically format test SVG <symbol>

* add basic <symbol> and tests

* Update changelog for SVG <symbol> support

* scale adjustment for symbols

* Fix <symbol> scaling using transforms instead of setattr

* Update step-security/harden-runner action to v2.19.2 (#1850)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* Update step-security/harden-runner action to v2.19.3 (#1851)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* Implement suggestions for <symbol> SVG support

---------

Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>
Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add Theo1335 as a contributor for code (#1853)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action action to v4.35.5 (#1852)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update crate-ci/typos action to v1.46.2 (#1854)

sthibaul committed:
markdown escaping: Fix injection of escapes (#1822)

c437630e1726 ("Fix markdown character escaping bug (issue #1236)")

fixed the escaping of multiple markdown markers, but it dropped dividing
by two the number of escaped characters being injected:

-                        txt_frag[: -((num_escape_chars + 1) // 2)]

+                    for _ in range(escape_run - 1):

This restores dividing by two the number of escapes, and restores the
test results accordingly, making it the behavior closer to
https://spec.commonmark.org/dingus/

Fixes: #1215

Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>

renovate[bot] committed:
Update github/codeql-action action to v4.36.0 (#1856)

renovate[bot] committed:
Update step-security/harden-runner action to v2.19.4 (#1855)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

andersonhc committed:
fix rebinding issue (#1857)

renovate[bot] committed:
Update crate-ci/typos action to v1.46.3 (#1858)

renovate[bot] committed:
Update crate-ci/typos action to v1.47.0 (#1859)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action action to v4.36.1 (#1860)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update actions/checkout action to v6.0.3 (#1861)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update crate-ci/typos action to v1.47.1 (#1862)

renovate[bot] committed:
Update crate-ci/typos action to v1.47.2 (#1863)

renovate[bot] committed:
Update github/codeql-action action to v4.36.2 (#1864)

andersonhc committed:
Fix actions workflow errors (#1867)

* add pypdf as test dependency

* fix errors test assertion

eugen-goebel committed:
Preserve leading spaces on new lines in <pre> blocks (#1866)

Fixes #1063.

The skip_leading_spaces default of Paragraph and
ParagraphCollectorMixin.paragraph() was False, so an explicitly passed
False got swallowed by the `or self._region.skip_leading_spaces`
fallback whenever the region had skip_leading_spaces=True, which is how
the HTML parser configures it. As a result, leading whitespace inside
<pre> and <pre><code> blocks was dropped.

The default is now None, and an explicit True or False always wins over
the region default, resolved once when the paragraph is constructed.
build_lines() and the bullet renderer use the paragraph's own resolved
value instead of falling back to the region again.

In html.py, _new_paragraph() now passes
skip_leading_spaces=not self._pre_formatted, so preformatted blocks keep
their leading whitespace while normal blocks behave as before.

A regression test based on the issue's minimal example is added in
test/html/.

Co-authored-by: Eugen Goebel <eugen-goebel@users.noreply.github.com>
Co-authored-by: Anderson Herzogenrath da Costa <andersonhc@gmail.com>

allcontributors[bot] committed:
add eugen-goebel as a contributor for code (#1868)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

renovate[bot] committed:
Update actions/checkout action to v7 (#1869)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update actions/setup-python action to v6.3.0 (#1871)

allcontributors[bot] committed:
add gaoflow as a contributor for code (#1873)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

gaoflow committed:
fix: set_font() early-return skips current_font restore when font state diverges (#1872)

* fix: set_font() early-return skips current_font restore when state diverges

eugen-goebel committed:
Add optional_content() context manager for screen-only / print-only content (#1870)

andersonhc committed:
Fix cff font (#1875)

renovate[bot] committed:
Update crate-ci/typos action to v1.48.0 (#1878)

Pawansingh3889 committed:
fix(text_region): preserve font state across render, closes #1804 (#1823)

allcontributors[bot] committed:
add iamfazakb as a contributor for doc (#1880)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action action to v4.36.3 (#1879)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

iamfazakb committed:
Fix broken anchor links (#1877)

* Fixed the broken anchor links for Margins.md, PageFormatAndOrientation.md, TextColumns.md and TextRegion.md.

* Fixed a broken anchor link in Images.md.

* Added an entry about fixing broken anchor links to CHANGELOG.md.

* Replaced links that lead to other parts of the documentation instead of the API reference.

renovate[bot] committed:
Update step-security/harden-runner action to v2.20.0 (#1881)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action action to v4.37.0 (#1882)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action action to v4.37.1 (#1886)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update pypa/gh-action-pypi-publish action to v1.14.1 (#1887)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update actions/setup-python action to v7 (#1888)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

uttam12331 committed:
Fix spurious 'Not enough horizontal space' with fragmented text (#1885)

MultiLineBreak.get_line guards against making no progress by comparing the
character index of the current forced break with the previous one. But the
character index is relative to the current fragment, and the guard did not
record which fragment the previous break happened in. With text that has no
break opportunities and is split into many small fragments (e.g. a fallback
font alternating with the main font on CJK text), consecutive lines can
legitimately break at the same character index within different fragments,
which raised a spurious FPDFException even though every line rendered
characters.

Track (fragment_index, character_index) for the last forced break so the
guard only fires when no progress was actually made. A single character
wider than the line still raises.

Fixes #1250

Co-authored-by: uttam12331 <uttam12331@users.noreply.github.com>

allcontributors[bot] committed:
add uttam12331 as a contributor for code (#1890)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

andersonhc committed:
Add SVG complexity limits (#1889)

* add SVGLimits

* add svg limit enforcement

* add documentation

* add changelog entries

* fix depth check

renovate[bot] committed:
Update actions/checkout action to v7.0.1 (#1891)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action action to v4.37.2 (#1896)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action action to v4.37.3 (#1897)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add dannymaaz as a contributor for code (#1898)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

dannymaaz committed:
Support SVG switch elements (#1892)

andersonhc committed:
bugfixes (#1900)

allcontributors[bot] committed:
add hvbtup as a contributor for bug (#1899)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

allcontributors[bot] committed:
add prateek-dagar as a contributor for code (#1901)

* update README.md [skip ci]

* update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

prateek-dagar committed:
feat: Allow keep_aspect_ratio flag for images in templates (#1893)

* Allow keep_aspect_ratio flag for images in templates

* Regenerate golden reference PDF for template keep_aspect_ratio test

renovate[bot] committed:
Update pypa/gh-action-pypi-publish action to v1.14.2 (#1906)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update github/codeql-action action to v4.37.4 (#1907)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
